File: night_watch/event_recorder.py
"""
Night Watch — Event Recorder (Sections 3.5–3.8)

Manages the lifecycle of acoustic events:
  1. Maintains a pre-event circular buffer (5 s of raw audio).
  2. Detects event start (N consecutive anomalous frames).
  3. Accumulates event audio while the anomaly persists.
  4. Detects event end (cooldown period of normal frames).
  5. Appends post-buffer audio.
  6. Writes the assembled clip as a .wav file.
  7. Records the event in the SQLite database.
  8. Merges events that are too close together (Phase 4).
"""

import os
import wave
from datetime import datetime, timedelta

import numpy as np

from night_watch.audio_buffer import CircularAudioBuffer
from night_watch.config import (
    COOLDOWN_FRAMES,
    MERGE_GAP_S,
    POST_BUFFER_S,
    PRE_BUFFER_S,
    SAMPLE_RATE,
    HOP_DURATION_S,
    TRIGGER_CONSECUTIVE,
)
from night_watch.database import NightWatchDB


class EventRecorder:
    """
    Call ``process_chunk()`` once per hop with the raw audio chunk and
    the detector's anomaly result.  The recorder handles everything else.

    Parameters
    ----------
    db : NightWatchDB
    session_id : int
    output_dir : str
        Directory to write .wav clips into.
    """

    # Event states
    IDLE = "idle"
    RECORDING = "recording"            # actively recording an event
    COOLING_DOWN = "cooling_down"      # anomaly stopped, counting cooldown
    POST_BUFFERING = "post_buffering"  # capturing post-event audio
    MERGE_WINDOW = "merge_window"      # waiting to see if another anomaly merges

    def __init__(self, db: NightWatchDB, session_id: int, output_dir: str):
        self.db = db
        self.session_id = session_id
        self.output_dir = output_dir
        os.makedirs(output_dir, exist_ok=True)

        # Pre-event buffer: always running
        pre_samples = int(PRE_BUFFER_S * SAMPLE_RATE)
        self._pre_buffer = CircularAudioBuffer(pre_samples)

        # State machine
        self._state = self.IDLE
        self._consecutive_anomalies = 0
        self._cooldown_remaining = 0
        self._post_remaining = 0
        self._merge_remaining = 0

        # Current event data
        self._event_audio: list[np.ndarray] = []
        self._event_start_time: datetime | None = None
        self._event_peak_score: float = 0.0

        # Event counter for this session
        self._event_count = 0

        # Completed events log (for display)
        self.completed_events: list[dict] = []

    def process_chunk(self, chunk: np.ndarray, is_anomaly: bool, score: float) -> dict | None:
        """
        Feed one hop of raw audio + detection result.

        Returns
        -------
        dict or None
            If an event just finished saving, returns its metadata.
            Otherwise None.
        """
        now = datetime.now()
        finished_event = None

        if self._state == self.IDLE:
            self._pre_buffer.write(chunk)
            if is_anomaly:
                self._consecutive_anomalies += 1
                if self._consecutive_anomalies >= TRIGGER_CONSECUTIVE:
                    self._start_event(now, score)
            else:
                self._consecutive_anomalies = 0

        elif self._state == self.RECORDING:
            self._event_audio.append(chunk.copy())
            self._event_peak_score = max(self._event_peak_score, score)
            if not is_anomaly:
                self._cooldown_remaining = COOLDOWN_FRAMES
                self._state = self.COOLING_DOWN

        elif self._state == self.COOLING_DOWN:
            self._event_audio.append(chunk.copy())
            self._event_peak_score = max(self._event_peak_score, score)
            if is_anomaly:
                # Anomaly resumed — keep recording
                self._state = self.RECORDING
                self._cooldown_remaining = 0
            else:
                self._cooldown_remaining -= 1
                if self._cooldown_remaining <= 0:
                    # Cooldown expired — start post-buffer
                    post_frames = int(POST_BUFFER_S / HOP_DURATION_S)
                    self._post_remaining = post_frames
                    self._state = self.POST_BUFFERING

        elif self._state == self.POST_BUFFERING:
            self._event_audio.append(chunk.copy())
            self._post_remaining -= 1
            if self._post_remaining <= 0:
                # Post-buffer done — enter merge window instead of saving
                merge_frames = int(MERGE_GAP_S / HOP_DURATION_S)
                self._merge_remaining = merge_frames
                self._state = self.MERGE_WINDOW

        elif self._state == self.MERGE_WINDOW:
            # Keep accumulating audio (gap audio preserved if events merge)
            self._event_audio.append(chunk.copy())
            if is_anomaly:
                # New anomaly during merge window → merge into same event
                self._state = self.RECORDING
                self._merge_remaining = 0
                self._event_peak_score = max(self._event_peak_score, score)
            else:
                self._merge_remaining -= 1
                if self._merge_remaining <= 0:
                    # Merge window expired — no new anomaly, save the event.
                    # Trim the merge window audio (it was just silence/ambient)
                    merge_samples = int(MERGE_GAP_S * SAMPLE_RATE)
                    finished_event = self._end_event(now, trim_tail=merge_samples)

        return finished_event

    def _start_event(self, now: datetime, score: float) -> None:
        """Transition from IDLE → RECORDING."""
        self._event_start_time = now
        self._event_peak_score = score
        self._event_audio = []

        # Grab the pre-buffer
        pre_audio = self._pre_buffer.read_all()
        if len(pre_audio) > 0:
            self._event_audio.append(pre_audio)

        self._state = self.RECORDING
        self._consecutive_anomalies = 0

    def _end_event(self, now: datetime, trim_tail: int = 0) -> dict:
        """Finalize and save the current event."""
        self._event_count += 1

        # Assemble the full clip
        if self._event_audio:
            full_audio = np.concatenate(self._event_audio)
        else:
            full_audio = np.array([], dtype=np.float32)

        # Trim merge-window silence from the tail if not merged
        if trim_tail > 0 and len(full_audio) > trim_tail:
            full_audio = full_audio[:-trim_tail]

        duration = len(full_audio) / SAMPLE_RATE

        # Save .wav
        filename = f"event_{self._event_count:04d}.wav"
        clip_path = os.path.join(self.output_dir, filename)
        self._write_wav(clip_path, full_audio)

        # Record in database
        start_str = self._event_start_time.isoformat() if self._event_start_time else now.isoformat()
        end_str = now.isoformat()

        event_id = self.db.insert_event(
            session_id=self.session_id,
            start_time=start_str,
            end_time=end_str,
            duration_seconds=round(duration, 2),
            peak_score=round(self._event_peak_score, 2),
            clip_path=clip_path,
        )

        event_info = {
            "id": event_id,
            "number": self._event_count,
            "start_time": self._event_start_time,
            "end_time": now,
            "duration": duration,
            "peak_score": self._event_peak_score,
            "clip_path": clip_path,
        }
        self.completed_events.append(event_info)

        # Reset for next event
        self._event_audio = []
        self._event_start_time = None
        self._event_peak_score = 0.0
        self._state = self.IDLE
        self._pre_buffer.clear()

        return event_info

    def finalize(self) -> None:
        """Call at session end — save any in-progress event."""
        if self._state in (self.RECORDING, self.COOLING_DOWN,
                           self.POST_BUFFERING, self.MERGE_WINDOW):
            self._end_event(datetime.now())

    @property
    def state(self) -> str:
        return self._state

    @property
    def event_count(self) -> int:
        return self._event_count

    # ── WAV writer ──────────────────────────────────────────────────────────

    @staticmethod
    def _write_wav(path: str, audio: np.ndarray) -> None:
        """Write float32 mono audio to a 16-bit PCM .wav file."""
        # Convert float32 [-1, 1] to int16
        audio_clamped = np.clip(audio, -1.0, 1.0)
        pcm = (audio_clamped * 32767).astype(np.int16)

        with wave.open(path, "wb") as wf:
            wf.setnchannels(1)
            wf.setsampwidth(2)  # 16-bit
            wf.setframerate(SAMPLE_RATE)
            wf.writeframes(pcm.tobytes())

File: night_watch_flutter/lib/core/recorder/event_recorder.dart
import 'dart:typed_data';

import 'package:path/path.dart' as p;

import 'package:night_watch_flutter/core/config.dart';
import 'package:night_watch_flutter/core/recorder/circular_buffer.dart';
import 'package:night_watch_flutter/core/recorder/wav_encoder.dart';
import 'package:night_watch_flutter/models/event_model.dart';
import 'package:night_watch_flutter/services/database_service.dart';

enum RecorderState {
  idle,
  recording,
  coolingDown,
  postBuffering,
  mergeWindow,
}

/// Event Recorder state machine managing pre-buffer, anomaly capture, cooldown,
/// post-buffering, event merging (Phase 4), and WAV encoding.
///
/// Ported from Python event_recorder.py.
class EventRecorder {
  final DatabaseService db;
  final int sessionId;
  final String outputDir;

  late final CircularAudioBuffer _preBuffer;

  RecorderState _state = RecorderState.idle;
  int _consecutiveAnomalies = 0;
  int _cooldownRemaining = 0;
  int _postRemaining = 0;
  int _mergeRemaining = 0;

  final List<Float32List> _eventAudio = [];
  DateTime? _eventStartTime;
  double _eventPeakScore = 0.0;
  int _eventCount = 0;

  final List<RecordedEvent> completedEvents = [];

  EventRecorder({
    required this.db,
    required this.sessionId,
    required this.outputDir,
  }) {
    final preSamples = (NightWatchConfig.preBufferS * NightWatchConfig.sampleRate).round();
    _preBuffer = CircularAudioBuffer(preSamples);
  }

  RecorderState get state => _state;
  int get eventCount => _eventCount;
  double get currentPeakScore => _eventPeakScore;

  /// Process one hop chunk of raw audio and detection results.
  Future<RecordedEvent?> processChunk({
    required Float32List chunk,
    required bool isAnomaly,
    required double score,
    DateTime? timestamp,
  }) async {
    final now = timestamp ?? DateTime.now();
    RecordedEvent? finishedEvent;

    switch (_state) {
      case RecorderState.idle:
        _preBuffer.write(chunk);
        if (isAnomaly) {
          _consecutiveAnomalies++;
          if (_consecutiveAnomalies >= NightWatchConfig.triggerConsecutive) {
            _startEvent(now, score);
          }
        } else {
          _consecutiveAnomalies = 0;
        }
        break;

      case RecorderState.recording:
        _eventAudio.add(Float32List.fromList(chunk));
        if (score > _eventPeakScore) _eventPeakScore = score;
        if (!isAnomaly) {
          _cooldownRemaining = NightWatchConfig.cooldownFrames;
          _state = RecorderState.coolingDown;
        }
        break;

      case RecorderState.coolingDown:
        _eventAudio.add(Float32List.fromList(chunk));
        if (score > _eventPeakScore) _eventPeakScore = score;
        if (isAnomaly) {
          _state = RecorderState.recording;
          _cooldownRemaining = 0;
        } else {
          _cooldownRemaining--;
          if (_cooldownRemaining <= 0) {
            final postFrames =
                (NightWatchConfig.postBufferS / NightWatchConfig.hopDurationS).round();
            _postRemaining = postFrames;
            _state = RecorderState.postBuffering;
          }
        }
        break;

      case RecorderState.postBuffering:
        _eventAudio.add(Float32List.fromList(chunk));
        _postRemaining--;
        if (_postRemaining <= 0) {
          final mergeFrames =
              (NightWatchConfig.mergeGapS / NightWatchConfig.hopDurationS).round();
          _mergeRemaining = mergeFrames;
          _state = RecorderState.mergeWindow;
        }
        break;

      case RecorderState.mergeWindow:
        _eventAudio.add(Float32List.fromList(chunk));
        if (isAnomaly) {
          // New anomaly during merge window -> merge seamlessly into same event
          _state = RecorderState.recording;
          _mergeRemaining = 0;
          if (score > _eventPeakScore) _eventPeakScore = score;
        } else {
          _mergeRemaining--;
          if (_mergeRemaining <= 0) {
            // Merge window expired -> trim silence tail and finalize
            final mergeSamples =
                (NightWatchConfig.mergeGapS * NightWatchConfig.sampleRate).round();
            finishedEvent = await _endEvent(now, trimTailSamples: mergeSamples);
          }
        }
        break;
    }

    return finishedEvent;
  }

  void _startEvent(DateTime now, double score) {
    _eventStartTime = now;
    _eventPeakScore = score;
    _eventAudio.clear();

    final preAudio = _preBuffer.readAll();
    if (preAudio.isNotEmpty) {
      _eventAudio.add(preAudio);
    }

    _state = RecorderState.recording;
    _consecutiveAnomalies = 0;
  }

  Future<RecordedEvent> _endEvent(DateTime now, {int trimTailSamples = 0}) async {
    _eventCount++;

    // Concatenate all audio chunks
    int totalLen = 0;
    for (final c in _eventAudio) {
      totalLen += c.length;
    }

    Float32List fullAudio;
    if (totalLen > 0) {
      fullAudio = Float32List(totalLen);
      int offset = 0;
      for (final c in _eventAudio) {
        fullAudio.setRange(offset, offset + c.length, c);
        offset += c.length;
      }
    } else {
      fullAudio = Float32List(0);
    }

    // Trim merge window silence from tail if not merged
    if (trimTailSamples > 0 && fullAudio.length > trimTailSamples) {
      fullAudio = fullAudio.sublist(0, fullAudio.length - trimTailSamples);
    }

    final duration = fullAudio.length / NightWatchConfig.sampleRate;
    final filename = 'event_${_eventCount.toString().padLeft(4, '0')}.wav';
    final clipPath = p.join(outputDir, filename);

    // Save 16-bit WAV file
    await WavEncoder.writeWavFile(
      filePath: clipPath,
      audio: fullAudio,
      sampleRate: NightWatchConfig.sampleRate,
    );

    final start = _eventStartTime ?? now;
    final event = RecordedEvent(
      sessionId: sessionId,
      eventNumber: _eventCount,
      startTime: start,
      endTime: now,
      durationSeconds: double.parse(duration.toStringAsFixed(2)),
      peakScore: double.parse(_eventPeakScore.toStringAsFixed(2)),
      clipPath: clipPath,
    );

    final eventId = await db.insertEvent(event);
    final savedEvent = event.copyWith(id: eventId);
    completedEvents.add(savedEvent);

    // Reset state for next event
    _eventAudio.clear();
    _eventStartTime = null;
    _eventPeakScore = 0.0;
    _state = RecorderState.idle;
    _preBuffer.clear();

    return savedEvent;
  }

  /// Finalize in-progress event on session stop.
  Future<RecordedEvent?> finalize() async {
    if (_state != RecorderState.idle && _eventAudio.isNotEmpty) {
      return await _endEvent(DateTime.now());
    }
    return null;
  }
}
